src/Core/Workspace/Workspace.cs

```diff
@@ -77,10 +88,74 @@ public string CacheFolder
+                if (SkipAutoLoadProject || projectFiles.Count() != 1) return Enumerable.Empty<string>();
+
+                var projectFilesToResolve = projectFiles.ToArray();
+                while (projectFilesToResolve.Any())
+                {
+                    try
+                    {
+                        Logger?.LogInformation($"Looking for project references in .csproj files: {string.Join(";", projectFilesToResolve)}");
+
+                        projectFilesToResolve = projectFilesToResolve
+                            .SelectMany(projectFile =>
+                                XDocument.Load(projectFile)
+                                    .XPathSelectElements("//ProjectReference")
+                                    .Select(element => Path.Combine(
+                                        Path.GetDirectoryName(projectFile),
+                                        element.Attribute("Include").Value.Replace('\\', Path.DirectorySeparatorChar))))
+                            .Where(projectFile => !projectFiles.Contains(projectFile))
+                            .ToArray();
+
+                        var missingFiles = projectFilesToResolve.Where(projectFile => !File.Exists(projectFile));
+                        if (missingFiles.Any())
+                        {
+                            Logger?.LogError($"Skipping invalid project references: {string.Join(";", missingFiles)}");
+                            projectFilesToResolve = projectFilesToResolve.Except(missingFiles).ToArray();
+                        }
+
+                        Logger?.LogInformation($"Adding project references: {string.Join(";", projectFilesToResolve)}");
+
+                        projectFiles = projectFiles.Concat(projectFilesToResolve);
+                    }
+                    catch (Exception e)
+                    {
+                        Logger?.LogError(e, $"Failed to resolve all project references: {e.Message}");
+                        projectFilesToResolve = Array.Empty<string>();
+                    }
+                }
+
+                return projectFiles.Distinct();
+            }
+        }
+
+        /// <summary>
+        /// Gets the source folders containing source files to be built for this Workspace.
+        /// </summary>
+        private IEnumerable<string> SourceFolders =>
+            ProjectFiles
+                .Select(projectFile => Path.GetDirectoryName(projectFile))
+                .Append(Root)
+                .Distinct();
+
         /// <summary>
         /// Gets the source files to be built for this Workspace.
         /// </summary>
-        public IEnumerable<string> SourceFiles => Directory.EnumerateFiles(Root, "*.qs", SearchOption.TopDirectoryOnly);
+        public IEnumerable<string> SourceFiles =>
+            SourceFolders.SelectMany(folder => Directory.EnumerateFiles(folder, "*.qs", SearchOption.TopDirectoryOnly));
```

Comment:
I think this makes sense given that we currently have a `<QsCompile Include="**/*.qs"/>` element by default, but as a future enhancement it may be good to resolve logical projects with MSBuild to handle when that implicit compile item is modified in a particular project.

Reply:
Yes, this is a good point. Using MSBuild to resolve this will be more robust. And this also points out another inconsistency here, since I'm using `SearchOption.TopDirectoryOnly`, but the default include directive `**/*.qs` would find source files in all subdirectories.